### ISSUE TITLE
provider-local: failed to create snapstore from configured storage provider: mkdir <omitted>: operation not permitted

### ISSUE THREAD

**Issue by ialidzhikov** [gardener/gardener#7236]
**How to categorize this issue?**

/area dev-productivity
/kind bug

**What happened**:
I am following https://github.com/gardener/gardener/blob/master/docs/development/getting_started_locally.md.
I created a local Shoot:
```
% kubectl create -f example/provider-local/shoot.yaml
```

The creation is stuck waiting etcd-main pod to come up:
```yaml
    lastErrors:
    - description: "task \"Waiting until main and event etcd report readiness\" failed:
        1 error occurred:\n\t* Error while waiting for Etcd shoot--local--local/etcd-main
        to become ready: retry failed with context deadline exceeded, last error:
        not enough ready replicas (0/1)\n\n"
      lastUpdateTime: "2022-12-19T12:43:08Z"
      taskID: Waiting until main and event etcd report readines
```

```
% k -n shoot--local--local get po
NAME            READY   STATUS    RESTARTS   AGE
etcd-events-0   2/2     Running   0          27m
etcd-main-0     1/2     Running   0          6m8s
```

```
% k -n shoot--local--local logs etcd-main-0 backup-restore

time="2022-12-19T12:46:39Z" level=error msg="Failed initialization: error while restoring corrupt data: failed to create snapstore from configured storage provider: mkdir 59eb917e-0551-4654-ac02-69c3086be3c7/shoot--local--local--29262d84-5dd4-426b-aa9a-e97fee14dbe9: operation not permitted" actor=backup-restore-server
time="2022-12-19T12:46:39Z" level=info msg="Responding to status request with: Failed" actor=backup-restore-server
time="2022-12-19T12:46:39Z" level=info msg="Updating status from Failed to New" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Responding to status request with: New" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Received start initialization request." actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Updating status from New to Progress" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Setting status to : 503" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Validation failBelowRevision: " actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Validation mode: full" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=error msg="Failed initialization: error while restoring corrupt data: failed to create snapstore from configured storage provider: mkdir 59eb917e-0551-4654-ac02-69c3086be3c7/shoot--local--local--29262d84-5dd4-426b-aa9a-e97fee14dbe9: operation not permitted" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Responding to status request with: Failed" actor=backup-restore-server
time="2022-12-19T12:46:40Z" level=info msg="Updating status from Failed to New" actor=backup-restore-server
```

**What you expected to happen**:
No such issues.

**How to reproduce it (as minimally and precisely as possible)**:
See above.

**Anything else we need to know?**:
@plkokanov reported that he also hits this issue and it appeared after https://github.com/gardener/gardener/pull/6939 for him (cc @vpnachev).

**Environment**:

- Gardener version: 44c7c5613ed79a618900a5bfacd3d812224fb3e1
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

**Comment by danielfoehrKn** [comment]
Also just hit the same issue (Docker-for-mac 4.15, grpc-Fuse file sharing)

TLDR; I updated Docker-for-Mac to the latest version 4.15 and tried all different hostpath sharing options in Docker-for-mac (grpc-Fuse, VirtioFS, osxfs) but the issue persists.
I can reproduce the same behavior of the backup-and-restore  process, namely that running as `root` user in the mount ns of the container, I do not have the permission to create a directory under the bind-mounted directory `bb7ec114-db2b-46da-897a-b1a0d3026031/...`.

What I have not checked yet
- seccomp 

# Things that I checked without success
```
time="2023-01-05T09:11:08Z" level=info msg="Temporary directory /var/etcd/data/temp does not exist. Creating it..."
time="2023-01-05T09:11:08Z" level=error msg="Failed initialization: error while restoring corrupt data: failed to create snapstore from configured storage provider: mkdir bb7ec114-db2b-46da-897a-b1a0d3026031/shoot--local--local--78f80c60-69e4-4f66-81b1-8b59fdf9a064: operation not permitted" actor=backup-restore-server
```


The directory `ls /proc/1797638/root/var/etcd/data/temp/` in the mount namespace of the etcd-backup restored got created successfully.

But the snapstore directory was indeed not created 
```
ls /proc/1797638/root/bb7ec114-db2b-46da-897a-b1a0d3026031/ -la
total 4
drwxr-xr-x 2 root root   64 Jan  3 09:41 .
drwxr-xr-x 1 root root 4096 Jan  5 09:11 ..
```

The snapstore tries to create the directory with permission `700`, see [here](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/snapstore/local_snapstore.go#L39), which is the same permission as for the tmp directory (see [here](https://github.com/gardener/etcd-backup-restore/blob/f69cd4deac013ed0cd5a0949a488be376fe9e87c/pkg/snapstore/utils.go#L62))

The permissions on the rootfs for the backup-restore
- notice that the `tmp` directory has different permissions, where anyone can `rwx`, but the `bb..` directory has `0755/drwxr-xr-x` only allowing the owner to write
- the `bb..` directory has already been created two days prior to attempted shoot creation
```
$ ls -la
drwxr-xr-x   2 root root   64 Jan  3 09:41 bb7ec114-db2b-46da-897a-b1a0d3026031
drwxrwxrwt   2 root root 4096 Aug  9 08:58 tmp
```

I can reproduce the same permission problem trying to create the snapstore directory when in the etcd-backup-restore's mount ns 
- note that the backup-restore process runs as `root` user
```
etcd-main-0:/# mkdir bb7ec114-db2b-46da-897a-b1a0d3026031/shoot--local--local--78f80c60-69e4-4f66-81b1-8b59fdf9a064
mkdir: can't create directory 'bb7ec114-db2b-46da-897a-b1a0d3026031/shoot--local--local--78f80c60-69e4-4f66-81b1-8b59fdf9a064': Operation not permitted
```


Now the question is, why the `root` user cannot create a directory in a directory marked as owned by `root`.
 - see that the rootfs is mounted from a containerd snapshot as `rw`
 - interestingly, there is a FUSE fs mounted for the `/bb7ec114-db2b-46da-897a-b1a0d3026031` directory, but mounted also `rw`
 
```
etcd-main-0:/# mount
overlay on / type overlay (rw,relatime,lowerdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/350/fs:/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/349/fs:/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/342/fs,upperdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/351/fs,workdir=/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs/snapshots/351/work)
grpcfuse on /bb7ec114-db2b-46da-897a-b1a0d3026031 type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
```

That makes sense, my docker desktop uses grpcFuse for filesharing with the host
![image](https://user-images.githubusercontent.com/33809186/210764329-2ae418c7-b63e-47cc-9014-15086d856127.png)

Note the volume mounts for the etcd-backup-restore pod
```
volumeMounts:
// this is our FUSE mounted directory 
 168   │     - mountPath: bb7ec114-db2b-46da-897a-b1a0d3026031 
 169   │       name: host-storage 

```
Hence, the directory `/etc/gardener/local-backupbuckets/` is bind-mounted from the Docker VM into the container.
```
- hostPath:
 225   │       path: /etc/gardener/local-backupbuckets/bb7ec114-db2b-46da-897a-b1a0d3026031
 226   │       type: Directory
 227   │     name: host-storage
```
The directory on the VM's fs shows that it is mounted with grpcFuse.
```
$mount
grpcfuse on /etc/gardener/local-backupbuckets type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
```

On my local machine, the backup bucket directory is created by the local provider-extension running as a root process (see [here](https://github.com/gardener/gardener/blob/c3e2599f43a86565194e44066b28614b3b7262dd/pkg/provider-local/controller/backupbucket/actuator.go#L49))

The kind cluster used for the local setup is then configured to mount the path `./dev/local-backupbuckets` into the docker desktop VM at `/etc/gardener/local-backupbuckets`, see [here](https://github.com/gardener/gardener/blob/6d18b043a3852034e571ccc9f5e265db1923fd45/example/gardener-local/kind/cluster/templates/_extra_mounts.tpl#L11-L10).

## Reverting to osxfs and updating DOcker-for-mac to 4.15

The [Kind documentation states](https://kind.sigs.k8s.io/docs/user/configuration/)

> NOTE: If you are using Docker for Mac or Windows check that the hostPath is included in the Preferences -> Resources -> File Sharing.

So that's what I did, I allowed 
- /Users/<User>/go/src/github.com/gardener/gardener/dev/local-registry
- /Users/<User>/go/src/github.com/gardener/gardener/dev/local-backupbuckets
- /Users/<User>/go/src/github.com/gardener/gardener/example/gardener-local/controlplane

![image](https://user-images.githubusercontent.com/33809186/210776171-7c564fee-7933-4354-a1bd-0e394346569b.png)

However, it still did not work  (`permission denied` when creating the directory under the grpc-fuse mounted directory)
- after a restart of docker-for-desktop
- after recreation of the kind cluster (seems to be a permanent problem)

Changing to **VirtioFS** available in docker 4.15 caused my docker-for-mac to not start any more when host path sharing is enabled like above.

I switched back to `osxfs` host path sharing + allowed the above mentioned host paths.
As expected, the mount in the counter looks like this:
```
osxfs on /baa84810-1fb6-49c0-ae29-c5c8a137e5b2 type fuse.osxfs (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
```
However, again the below command  fails & so does the backup-and-restore

```
mkdir baa84810-1fb6-49c0-ae29-c5c8a137e5b2/shoot--local--local--25f779c5-ec14-4151-9dfa-142661f14509:
```

**Comment by danielfoehrKn** [comment]
**Result**: the permission of the directory on my local machines are incorrect and have been changed from `0775` to `0751` not allowing the `osxfs` process (FUSE) to write to the directory. Hence, all operations on the mounted bucket directory  from within the VM get a permission error.

# Further
Creating a file in the sibling directory of the mounted directory in the docker VM works and is visible from my local machine filesystem. 

```
 //Permission denied to create a new file in existing bucket directory
echo "hi" > /etc/gardener/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/gago.txt


// creating a new backup-bucket directory works
mkdir  /etc/gardener/local-backupbuckets/new_bucket

// creating a new file in the new backupbucket directory also works 
echo "hi" > /etc/gardener/local-backupbuckets/new_bucket/gago.txt
```


Permission wise, my new bucket directory looks identical to the actual bucket directory `baa84810-1fb6-49c0-ae29-c5c8a137e5b2`
- so that hints at a problem on my local machine rather than in the VM

```
root@gardener-local-control-plane:/# stat  /etc/gardener/local-backupbuckets/new_bucket
  File: /etc/gardener/local-backupbuckets/gago
  Size: 96        	Blocks: 0          IO Block: 4096   directory
Device: 90h/144d	Inode: 314240223   Links: 3
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-01-06 08:40:28.000000000 +0000
Modify: 2023-01-06 08:40:23.000000000 +0000
Change: 2023-01-06 08:40:23.000000000 +0000
 Birth: -

root@gardener-local-control-plane:/# stat  /etc/gardener/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/
  File: /etc/gardener/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/
  Size: 64        	Blocks: 0          IO Block: 4096   directory
Device: 90h/144d	Inode: 314050651   Links: 2
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-01-05 15:10:10.000000000 +0000
Modify: 2023-01-05 15:10:10.000000000 +0000
Change: 2023-01-05 15:10:10.000000000 +0000
 Birth: -
```


Well,  my privileged user on OSX is also not allowed to write to the local backupbucket directory 

```
echo "hello" > dev/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/hello.txt
zsh: permission denied: dev/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/hello.txt
```


checking the directories permission reveals that my privileged user did not create the directory, but the `root` user. That makes sense, as the `make start-local-provider` script starts the golang process with  `sudo`.

Writing to the local backup-bucket directory as `root` in OSX works.
- the file is also propagated to the VM's root fs as well as in the containers bind-mounted directory
 
```
bash-5.0# whoami
root
bash-5.0# echo "hello" > dev/local-backupbuckets/baa84810-1fb6-49c0-ae29-c5c8a137e5b2/hello.txt
```

Turns out the permission on that directory is `0751` or `rwxr-xr-x`, created by user `root` in group `staff`. Hence, only the owner can actually write to this directory - but not my personal user (no matter if it is in the same `staff` group).

Turns out, the `osxfs` FUSE (note: userspace) process runs as my local admin account, not as `root`. That should explain why writing to the OSX directory from within the VM does not work.

```
bash-5.0# ps -ej | grep osx
<my_osx_user>          14261  5813  5813      0    1 S      ??    1:16.53 com.docker.osxfs serve --address fd:3 --transfused transfused.sock --control fd:4
```

After changing the osx permissions to `0775`, writing from within the docker VM to the backup bucket directory worked. Also the logs in etcd-backup-restore indicate success

```
time="2023-01-06T09:42:55Z" level=info msg="Finding latest set of snapshot to recover from..."
time="2023-01-06T09:42:55Z" level=info msg="No snapshot found. Will remove the data directory."
time="2023-01-06T09:42:55Z" level=info msg="Removing directory(/var/etcd/data/new.etcd) since snapstore is empty."
time="2023-01-06T09:42:55Z" level=info msg="Successfully initialized data directory for etcd." actor=backup-restore-server
time="2023-01-06T09:42:55Z" level=info msg="Responding to status request with: Successful" actor=backup-restore-server
time="2023-01-06T09:42:55Z" level=info msg="Updating status from Successful to New" actor=backup-restore-server
```


**However**: I do not know why my OSX directory had permissions `0751` and not `0775` as it should have according to the [provider-locals code](https://github.com/gardener/gardener/blob/c3e2599f43a86565194e44066b28614b3b7262dd/pkg/provider-local/controller/backupbucket/actuator.go#L49) (the history has shown that the same permissions are used since it's inception).
